### ISSUE TITLE
Fixed amount discounts are not parsed in the configured locale.

### ISSUE THREAD

**Issue by lsahidin** [opensourcepos/opensourcepos#3114]
### Background information

__IMPORTANT: If you choose to ignore this issue report template, your issue will be closed as we cannot help without the requested information.__

Please make sure you tick (add an x between the square brackets with no spaces) the following check boxes:

- [] Reporting an issue of an unmodified OSPOS installation
- [x] Checked [open and closed issues](https://github.com/opensourcepos/opensourcepos/issues?utf8=%E2%9C%93&q=is%3Aissue) and no similar issue was already reported (please make sure you searched!)
- [x] Read [README](https://github.com/opensourcepos/opensourcepos/blob/master/README.md), [WHATS_NEW](https://github.com/opensourcepos/opensourcepos/blob/master/WHATS_NEW.txt), [INSTALL.md](https://github.com/opensourcepos/opensourcepos/blob/master/INSTALL.md) and [UPGRADE](https://github.com/opensourcepos/opensourcepos/blob/master/UPGRADE.txt)
- [x] Read the [FAQ](https://github.com/opensourcepos/opensourcepos#faq) for any known install and/or upgrade gotchas (in specific PHP extensions installed)
- [x] Read the [wiki](https://github.com/opensourcepos/opensourcepos/wiki)
- [] Executed any database upgrade scripts if an upgrade pre 3.0.0 (e.g. database/2.4_to_3.0.sql)
- [] Aware the installation code is in [bintray](https://bintray.com/jekkos/opensourcepos/opensourcepos/view/files?sort=updated&order=asc#files) (see README), and [GitHub master](https://github.com/opensourcepos/opensourcepos/tree/master) is for [developers only](https://github.com/opensourcepos/opensourcepos/wiki/Development-setup) and therefore not complete nor stable

### Installation information

- OSPOS version is: 3.3.3
- OSPOS git commit hash is: 
- PHP version is: 7.3.26
- MySQL or MariaDB version is: MySQL
- OS and version is: Windows 10
- WebServer is: Xampp
- Selected language is: 
- (If applicable) Docker installation: 
- (If applicable) Installation package for the LAMP/LEMP stack is: 

### Issue / Bug  / Question / New Feature

In Sales menu I add item by entering item number, then select customer. After that I try to increase qty but error occurs with msg "Error editing item" I am just thinking may be it cannot increase qty if customer is already selected. so I try to remove the customer then increase the qty, but same condition happen still with error msg "Error editing item". just to make sure that the item stock quantity is available. then I try to add item by entering same item number again, it success increase the qty +1 without error. So, what I miss.. is anybody have same issue?
![laporan-error](https://user-images.githubusercontent.com/12827784/108628256-73160b00-748c-11eb-8d44-1f1e385b24c9.PNG)

**Comment by odiea** [comment]
In Employee permissions do you have everything checked under Sales for correct permissions?

**Comment by lsahidin** [comment]
Yes, sure. I am using admin

**Comment by ozzylamberth** [comment]
It is a warning error that happens to me when I charge the Client first than the item to be sold, oh when the transaction is about to end I want to increase the quantity of the item. I just remove the sale and charge the item again.
That error only happens in those two ways. It works normally when you load the item, you increase or decrease the amount of it, then load the customer and finalize the sale

**Comment by lsahidin** [comment]
The problem is on discount validation, when I changed the Localization to this value "1.234.567.890,12" (not default localization code) the problem occurs, so to avoid error the discount value on every item in the cart while you update the quantity you need to set value without decimal "0.00" to "0" it will work without error. But when you set Localization back to default en_US with result format will be like "$1,234,567,890.12" discount with decimal "0.00" will work without problem. How to fix this problem?

**Comment by jekkos** [comment]
do you mean that the discount field is not rendered according to the chosen localization format and you need to edit it in order to make it work in that case?

**Comment by lsahidin** [comment]
yes sir, like as you mean discount field not rendered according to chosen localization format. thanks for your attention.

**Comment by jekkos** [comment]
I made a pull request for this, but still wonder how we want to render / parse fixed amount discounts

* For currency discounts, we just render / parse in the configured currency precision which is obvious
* For discount percentages, either we use the configured decimals for quantities, or we stick to 0 decimals

In the current version, I have set it to 0 decimals (as it was before). This means that percentage discounts will never rendered with any decimals. For my usecase this is fine as we always discount using rounded percentages. Despite this it's still possible to enter a decimal amount and it will be parsed using the currency decimal setting as a max precision

**Comment by jekkos** [comment]
Any feedback welcome here @daN4cat @objecttothis @SteveIreland

**Comment by jekkos** [comment]
@lsahidin any feedback on these changes would be welcomed.

**Comment by objecttothis** [comment]
@jekkos I don't know if this is what you are referring to, but I did notice that the cash discount is not rendering according to the localization.  If you enter the localization to something with a comma as the decimal separator then in the sale register you enter the discount amount using the proper localization, upon clicking the update button it renders the cash discount with a period as the decimal separator, but it should be a comma.  In the example below, I entered a discount of 14,15, but when you click update you can see it renders as 14.15.

![image](https://user-images.githubusercontent.com/17935339/113705447-7e489180-96ee-11eb-856d-dd88c118efb7.png)

As for rendering percentages, I think we should render the precision entered rather than the setting... this is because the setting, at least in English, is called "Cash Decimals" and "Quantity Decimals".  Those should apply only to cash and quantities.  IMO we should either have a third setting for percentage precision or use a standard precision.

As an aside, I think our naming for the various precision variables is confusing.  We have Currency Decimals and Cash Decimals.  I think Currency Decimals is how many decimal places are shown after the separator and cash decimals is the rounding precision.  Regardless of whether that is correct or not, I think we could benefit from a few more tooltip icons to give an explanation of what those settings do.

**Comment by SteveIreland** [comment]
If it's working correctly, Cash decimals is the rounding precision to be used only if paid in full using the payment type of Cash.

**Comment by jekkos** [comment]
@objecttothis regarding the rendering, I have skimmed through all views and we don't consistently use rounding on a discount percentage. I think your suggestion to render the precision a user has used for the entry makes sense.

However I'm not sure if its feasible. If we fix the first issue you described then we might need to stick to a fixed set of decimals for the locale configured rendering.

The parsing fix for this issue is to use the parse_decimals to parse the discount entry, so in case a user enters a locale specific delimiter, then the parser should not reject it. Of course the rendering should then happen in the same way otherwise things might become very confusing.

**Comment by jekkos** [comment]
@objecttothis I think I have found a fix for the rendering now. If you call to_decimals without any precision, it will render the amount with max 2 decimals. If no decimal is present, then none will be rendered.

Also I fixed the discount rendering in most of the templates. It wasn't rendered in the correct locale in case decimals are present.

**Comment by objecttothis** [comment]
@jekkos that seems as good a fix as any.

**Comment by jekkos** [comment]
@lsahidin can you give this a try and report back ? I have installed this version on dev.opensourcepos.org for your convenience.

**Comment by lsahidin** [comment]
@jekkos I have tried on dev.opensourcepos.org with my current configuration to produce error, but before that, why the default discount is add 2% on every item? I check on items there is no discount available. so, where is the 2% discount come from? maybe I miss something.

![Screenshot 2021-04-08 122026](https://user-images.githubusercontent.com/12827784/113972469-dc978080-9864-11eb-8801-c261ea882500.png)

**Comment by jekkos** [comment]
@lsahidin the default discount can be configured in the settings. Currently it's set to 2% indeed.

**Comment by jekkos** [comment]
@lsahidin can you confirm if the parsing works for your locale? I see you have set the language to id.

**Comment by lsahidin** [comment]
> @lsahidin can you confirm if the parsing works for your locale? I see you have set the language to id.

Yes, I try to configure with my locale to ID then it works for me with round discount percent. but won't work if discount in decimal 0.0 so it doesn't matter for me cause discount only in round value. no error occur when I try to using comma 0,5 but not affected to discount total.

**Comment by jekkos** [comment]
Indeed the parsing works now. I think there might still be an issue in the configuration, you need to increase the quantity decimals to have a decimal outcome on the price.

**Comment by jekkos** [comment]
It seems that I did forget to update the receivings receipt. Also I can't really say how rounding should be configured now. 
* If I set currency decimals to 2 and add a discount with decimals, the price does not have any decimals.

Maybe this precision for different inputs should be simplified as we after this change we parse a number with the precision that it is entered with. 

So you can still control the precision of the input by adding decimals. They will be taken along and the final result should then be rounded to the specified precision.

Does this make sense @SteveIreland @objecttothis ?

**Comment by SteveIreland** [comment]
I'll also check it out tonight.  The discount amount should be rendered in the currency decimals.  The cash decimals should only kick in when different from the currency decimals and the payment type is cash.  At least that is what was intended to happen.  So if the cash decimals is zero then what you described should be correct if the payment type was also cash.  But it shouldn't have anything to do with how the discount amount is entered.

**Comment by jekkos** [comment]
I have pushed one more fix for the discount rendering on the receivings receipt.

I was testing some combinations of quantity and currency decimals earlier. I would expect that if you add a discount with decimal, and have currency decimals also larger than 0, you would at least see one decimal on the price in case a decimal discount is applied. This didn't seem to be the case so maybe we can log it as a separate ticket.. Not sure if we want to fix this before 3.3.4 as it's already a bit  overdue.

**Comment by jekkos** [comment]
I might want to retest this before my changes. Will do so later tonight so we can hopefully close this soon.

**Comment by jekkos** [comment]
It seems that setting currency and cash decimals alone is not enough to have a decimal in the final price. I also need to set tax decimals before I see any decimals in the end.

Just think that maybe we should review these settings and keep on decimal setting only for the final price. 
We can parse all other input as they are entered. This should go into a new ticket though and I don't want to hold this release to fix it.

@lsahidin can you check if the decimal discount is applied properly if you set tax decimals to the same amount of decimals you want to use for the discount input?

Thanks.

**Comment by jekkos** [comment]
the build is still on dev so it can be verified there @odiea @lsahidin

**Comment by odiea** [comment]
I will give it a try later today. Thanks

**Comment by odiea** [comment]
I tried it out and it appears to work ok but I have not had issues with rounding.

**Comment by jekkos** [comment]
Ok so all fine? Let's merge then!

**Comment by jekkos** [comment]
Release 3.3.4 page is created. I'll attach the zip files once the build is ready.

**Comment by objecttothis** [comment]
@jekkos I'm getting a numberformatter error

Severity: error --> Exception: numfmt_create: number formatter creation failed /usr/local/www/public_html/pos_database/application/helpers/locale_helper.php 401

I get this error when loading the items page and when trying to load the localization settings.  It looks like you changed this section of code recently.  Are you able to recreate the issue?

**Comment by objecttothis** [comment]
Same with Severity: error --> Exception: numfmt_create: number formatter creation failed /usr/local/www/public_html/pos_database/application/controllers/Config.php 343

**Comment by objecttothis** [comment]
nevermind.  I restarted mysql and apache and the errors went away.  I think it may be related to a recent upgrade of apache that I did.  I thought it automatically restarted the service but it must not have.
